### PR TITLE
feat(auth): add updateMFAPreference API

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

`updateMFAPreference` API updates the current MFA settings options for the user. This API is meant to be a cognito specific and mainly used after the user is authenticated

**TODO List:**
- Get accessToken from auth token provider once implementation is complete

#### Description of how you validated changes

Created a sample app where I was able to update MFA settings for the user

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/__tests__/providers/cognito/updateMFAPreference.test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import { authAPITestParams } from './testUtils/authApiTestParams';
+import { AuthError } from '../../../src/errors/AuthError';
+import {
+	SetUserMFAPreferenceException,
+} from '../../../src/providers/cognito/types/errors';
+import { AmplifyErrorString } from '@aws-amplify/core';
+import { UpdateMFAPreferenceRequest } from '../../../src/providers/cognito/types';
+import { getMFASettings } from '../../../src/providers/cognito/apis/updateMFAPreference';
+
+const mfaChoises: UpdateMFAPreferenceRequest[] = [
```
Comment (jimblanc) [inline packages/auth/__tests__/providers/cognito/updateMFAPreference.test.ts:16]:
```suggestion
const mfaChoices: UpdateMFAPreferenceRequest[] = [
```

File: packages/auth/__tests__/providers/cognito/updateMFAPreference.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+			.spyOn(setUserMFAPreferenceClient, 'setUserMFAPreferenceClient')
+			.mockImplementationOnce(async () => {
+				return {} as SetUserMFAPreferenceCommandOutput;
+			});
+	});
+	afterEach(() => {
+		setUserMFAPreferenceClientSpy.mockClear();
+	});
+	test.each(mfaChoises)(
+		'setUserMFAPreferenceClient should be called with all possible mfa combinations',
```
Comment (jimblanc) [inline packages/auth/__tests__/providers/cognito/updateMFAPreference.test.ts:51]: Dig the thoroughness!

File: packages/auth/src/providers/cognito/apis/updateMFAPreference.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import { setUserMFAPreferenceClient } from '../utils/clients/SetUserMFAPreferenceClient';
+import { CognitoMFASettings } from '../utils/clients/types/models';
+
+/**
+ * Updates the MFA preference of the user.
+ *
+ * @param updateMFAPreferenceRequest - The request object to update MFA preference.
+ *
+ * @throws -{@link SetUserMFAPreferenceException } - Service error thrown when the MFA preference cannot be updated.
+ *
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:16]:
```suggestion
 * @throws {@link SetUserMFAPreferenceException}
 * Service error thrown when the MFA preference cannot be updated.
```

File: packages/auth/src/providers/cognito/apis/updateMFAPreference.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * TODO: add config errors
+ */
+export async function updateMFAPreference(
+	updateMFAPreferenceRequest: UpdateMFAPreferenceRequest
+): Promise<void> {
+	const { sms, totp } = updateMFAPreferenceRequest;
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:23]: Do we need input validation if both of these are undefined?

Comment (israx) [reply packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:23]: as part of the org design, we are relaying of service errors mainly

File: packages/auth/src/providers/cognito/apis/updateMFAPreference.ts (hunk 3 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:16]: the issue of not having a character after `@throws` is that the docs string won't catch the {@link SetUserMFAPreferenceException}

**Review by jimblanc — DISMISSED** [review]

File: packages/auth/src/providers/cognito/utils/clients/types/models.ts [hunk 5]
```diff
@@ -21,7 +21,12 @@ export type ChallengeParameters = {
...
 	PASSWORD_CLAIM_SIGNATURE?: string;
 	MFAS_CAN_CHOOSE?: string;
-	MFAS_CAN_SETUP?:string
+	MFAS_CAN_SETUP?: string;
 } & { [Params: string]: unknown };
 
 export type CognitoMFAType = 'SMS_MFA' | 'SOFTWARE_TOKEN_MFA';
+
+export type CognitoMFASettings = {
+	Enabled?: boolean;
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/types/models.ts:30]: Why is optional?

File: packages/auth/src/providers/cognito/utils/clients/HttpClients.ts [hunk 6]
```diff
@@ -20,6 +20,8 @@ import type {
 	VerifySoftwareTokenCommandOutput,
 	AssociateSoftwareTokenCommandInput,
 	AssociateSoftwareTokenCommandOutput,
+	SetUserMFAPreferenceCommandInput,
+	SetUserMFAPreferenceCommandOutput,
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/HttpClients.ts:24]: I would add A TODO here for using internal types

File: packages/auth/src/providers/cognito/apis/updateMFAPreference.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	if (mfaPreference === 'DISABLED') {
+		return {
+			Enabled: false,
+		};
+	} else if (mfaPreference === 'PREFERRED') {
+		return {
+			Enabled: true,
+			PreferredMfa: true,
+		};
+	} else if (mfaPreference === 'ENABLED') {
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:45]: What is the difference between this and `NOT_PREFERRED`?

**Review by elorzafe — COMMENTED** [review]
Overall looks great!

We can discuss the comments offline if you want

Thanks @israx

File: packages/auth/src/providers/cognito/utils/clients/types/models.ts (hunk 5 above)
Comment (israx) [reply packages/auth/src/providers/cognito/utils/clients/types/models.ts:30]: The underlying service doesn't require this param

File: packages/auth/src/providers/cognito/apis/updateMFAPreference.ts (hunk 7 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/updateMFAPreference.ts:45]: A customer would set up totp or/and sms as `ENABLED` when a MFA option is not preferred. The `NOT_PREFERRED` should be used when an MFA option was already selected as preferred.

**Review by elorzafe — APPROVED** [review]
Good stuff @israx 🏅

**Review by elorzafe — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Pull request merged**